### DISCUSSION TITLE
Insert pore space instead of a sphere/grain?

### DISCUSSION THREAD

**Discussion by akkendrick** [PMEAL/porespy#337] [Q&A] [answered]
I'm sorry if I missed this in the documentation, but is there an easy way currently in the code to insert a hole in a grain? I think I could do this with insert shape but there might be an easier way. My goal is to leverage porespy to systematically add a secondary/much smaller porosity to an existing segmented image for reference.

**Comment by jgostick** [comment, marked as answer]
First of all, I'm loving that this is a 'discussion' rather than an 'issue'. I will attempt to answer.

It sound like you have a tomography image of large features (i.e. rock grains) and you wan to add holes to the grains, or something like that.  

You have a few options:

1.  You can generate a image with much smaller features than your grains, then multiplying the images together...being sure to 'swap' phases as needed, so that you set void to solid (e.g., ``im = ~im``), then multiply, then revert back.  Below is the basic idea and result:

```
import porespy as ps
import matplotlib.pyplot as plt

im = ps.generators.blobs([400, 400])
im2 = ps.generators.blobs(shape=[400, 400], blobiness=3)
im3 = im * im2
fig, ax = plt.subplots(1, 3)
ax[0].imshow(im)
ax[1].imshow(im2)
ax[2].imshow(im3)
```
![blobs](https://user-images.githubusercontent.com/5092049/103579413-455a1000-4ea6-11eb-8041-124039c26258.png)

2. You could use RSA, which takes an input image (your solid grains, with solid set to 0) and inserts spheres into the 0 regions until a specified volume fraction is reached.  

```
im = ps.generators.blobs([400, 400])
im2 = ps.generators.RSA(im=im, radius=5)
fig, ax = plt.subplots(1, 2)
ax[0].imshow(im)
ax[1].imshow(im2)
```
![spheres](https://user-images.githubusercontent.com/5092049/103579681-bb5e7700-4ea6-11eb-9e12-fc25d0bb21ec.png)

**Reply by akkendrick** [reply]
Yes that is exactly right thanks! It looks like the first approach should definitely work, I just had to sort some boolean/int issues on my end.

**Reply by jgostick** [reply]
glad to help.  be sure to mark my answer as correct :-)  so your question shows up as solved for future generations.
